### DISCUSSION TITLE
Deploying on Anvil is not working

### DISCUSSION THREAD

**Discussion by recepcankaya** [Cyfrin/foundry-full-course-cu#866] [Q&A]
When I try to deploy the script to Anvil, it gives the following error
```
[⠆] Compiling...
No files changed, compilation skipped
Script ran successfully.

EIP-3855 is not supported in one or more of the RPCs used.
Unsupported Chain IDs: 31337.
Contracts deployed with a Solidity version equal or higher than 0.8.20 might not work properly.
For more information, please see https://eips.ethereum.org/EIPS/eip-3855

## Setting up (1) EVMs.

==========================

Chain 31337

Estimated gas price: 5 gwei

Estimated total gas used for script: 802075

Estimated amount required: 0.004010375 ETH

==========================

###
Finding wallets for all the necessary addresses...

Transactions saved to: /home/recepcnky/foundry-tutorial/erc20/broadcast/DeployOurToken.s.sol/31337/run-latest.json

Sensitive values saved to: /home/recepcnky/foundry-tutorial/erc20/cache/DeployOurToken.s.sol/31337/run-latest.json

Error: 

You seem to be using Foundry's default sender. Be sure to set your own --sender.
No associated wallet for addresses: {0x1804c8ab1f12e6bbf3894d4083f33e07309d1f38}. Unlocked wallets: []
```
I think, it successfully deploys but somehow returns an error. By the way, the normal deploy works without passing --broadcast. The issue starts when I try to simulate on-chain transactions. I saw some issues opened in discussion related with this, but I could not see a proper solution.

**Comment by 0xusmanf** [comment]
When you simulate on-chain transactions you have to provide your private key of the account you want to originate the transactions from, for example if you are trying to simulate on anvil then you can provide one of the anvil's default account's private key like below.
```shell
forge script script/DeployFundMe.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
```

After providing the private key it should work.

**Reply by sadityakumar9211** [reply]
Check if you're providing the --broadcast flag in the forge command. I was missing that. After adding `--broadcast` flag, deploying on anvil is working for me, although I am getting the same warning message.

**Comment by azanux** [comment]
hi @recepcankaya 

You could now use cast **wallet import** to save and encrypt your private key : [[cast wallet](https://book.getfoundry.sh/reference/cast/wallet-commands)](https://book.getfoundry.sh/reference/cast/wallet-commands)

Here are how you can use it  :
1 - import your private key into cast wallet , you will have to paste your private key and set a password to encrypt it , replace **defaultKey** with the name of your private key like (devKey, devPk, prodKey ...)
`cast wallet import **defaultKey** —-interactive`

2 - Now you can run your script to deploy your contract, it will ask your password before sending the transaction ( replace defaultKey with the name you set before and replace the publickey by the publickey of your wallet)
`forge script script/DeployFundMe.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --sender **0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266** --account **defaultKey**`

**Reply by biadebeatriz** [reply]
Works for me!!!!

**Reply by EngrPips** [reply]
Indeed

**Comment by abhithshaji** [comment]
Had this error when I was using the private key from keystore, ie, after I set up the private keys using command : 
` cast wallet import --interactive testAccount`.
and then when I used the command : 
`forge script script/DeploySimpleStorage.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --account testAccount `.

**Solution** : Just added the address associated with the private key using `--sender`.

eg : `forge script script/DeploySimpleStorage.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --account testAccount  --sender 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`

**Reply by azanux** [reply]
@Abhith-S   Hello, actually if you read point 2 of my response carefully, there was already the 'sender' attribute in the command (just after --broadcast) , you don't need to specify the 'password' attribute in your command either, because it will ask for it afterwards. (unless you want to run it from a make file, in which case you need to add the password in a .env file, so you don't have to show your password).

**Reply by abhithshaji** [reply]
Oh yes. I kinda missed that. 
And thanks for the tip @azanux .

**Reply by roudra323** [reply]
I made the changes but still getting the warning.
```bash
roudra323@Roudra:/home/Cyfrin_tutorials/foundry/foundry_2$ forge script script/DeploySimpleStorage.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --account testAccount  --sender 0x70997970C51812dc3A010C7d01b50e0d17dc79C8
[⠒] Compiling...
No files changed, compilation skipped
EIP-3855 is not supported in one or more of the RPCs used.
Unsupported Chain IDs: 31337.
Contracts deployed with a Solidity version equal or higher than 0.8.20 might not work properly.
For more information, please see https://eips.ethereum.org/EIPS/eip-3855
Script ran successfully.

== Return ==
0: contract SimpleStorage 0x71C95911E9a5D330f4D621842EC243EE1343292e

## Setting up 1 EVM.

==========================

Chain 31337

Estimated gas price: 4.188392894 gwei

Estimated total gas used for script: 464097

Estimated amount required: 0.001943820576926718 ETH

==========================
Enter keystore password:
##
Sending transactions [0 - 0].
⠁ [00:00:00] [####################################################################################################################] 1/1 txes (0.0s)##
Waiting for receipts.
⠉ [00:00:00] [################################################################################################################] 1/1 receipts (0.0s)
##### anvil-hardhat
✅  [Success]Hash: 0x0f8fd4434cbadf999953dcf54bffe294616b4a2cd9bd53ac5a6eba4a6cb213dd
Contract Address: 0x71C95911E9a5D330f4D621842EC243EE1343292e
Block: 6
Paid: 0.001257553260859456 ETH (357088 gas * 3.521690062 gwei)



==========================

ONCHAIN EXECUTION COMPLETE & SUCCESSFUL.
Total Paid: 0.001257553260859456 ETH (357088 gas * avg 3.521690062 gwei)

Transactions saved to: /home/Cyfrin_tutorials/foundry/foundry_2/broadcast/DeploySimpleStorage.s.sol/31337/run-latest.json

Sensitive values saved to: /home/Cyfrin_tutorials/foundry/foundry_2/cache/DeploySimpleStorage.s.sol/31337/run-latest.json
```

what can be the reason ?

**Reply by vigneshbalu96** [reply]
It seems your code was successfully deployed and the top warnings are like instructions, you can ignore that! @roudra323

**Reply by vigneshbalu96** [reply]
you can use the commands as below :point_down: 

For keystore encryption using password :point_right: `cast wallet import testAccount --interactive`

Above command asks to enter the private key of the associated account that you are using to deploy the contract and also the password to access that private key in the keystore (These two are invisible fields after entering(private key & password) values for the keystore)

Deployment Script :point_right: ```forge script script/DeployFundMe.s.sol:DeployFundMe --rpc-url http://127.0.0.1:8545 --broadcast --sender 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 --account testAccount```

The above command asks for the password to the keystore to safely access the private key which we set before.

After Entering the password this would work as expected!!!

**Reply by EngrPips** [reply]
Interesting.
